File: src/eu/shooktea/passkeeper/Storage.scala
package eu.shooktea.passkeeper

import javafx.beans.property.SimpleListProperty
import javafx.collections.{FXCollections, ObservableList}

object Storage {

  def savedCipherableElement(element: Cipherable[_]) : Unit = {
    if (currentlyModifiedIndex == -1)
      store(element)
    else
      update(currentlyModifiedIndex, element)
    currentlyModifiedIndex = -1
  }

  def store(element: Cipherable[_]) : Unit = {
    elements add element
  }

  def update(index: Int, element: Cipherable[_]) : Unit = {
    elements set(index, element)
  }

  def remove(index: Int) : Unit = {
    elements remove index
  }

  def getElements() : ObservableList[Cipherable[_]] = elements.get()
  def setElements(list: ObservableList[Cipherable[_]]) : Unit = elements set list
  def elementsProperty() : SimpleListProperty[Cipherable[_]] = elements

  def ++(element: Cipherable[_]) : Unit = store(element)
  def --(index: Int) : Unit = remove(index)

  private val elements: SimpleListProperty[Cipherable[_]] = new SimpleListProperty(FXCollections.observableArrayList[Cipherable[_]]())
  private var currentlyModifiedIndex: Int = -1
}

File: src/eu/shooktea/passkeeper/Storage.java
package eu.shooktea.passkeeper;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Storage {
    private Storage() {}

    public static void saveCipherableElement(Cipherable c) {
        if (editedIndex == -1) {
            store(c);
        }
        else {
            update(editedIndex, c);
        }
        editedIndex = -1;
    }

    public static void clear() {
        elements.clear();
    }

    public static void store(Cipherable c) {
        elements.add(c);
    }

    public static void update(int index, Cipherable c) {
        elements.set(index, c);
    }

    public static <T extends Cipherable> List<T> filter(Type t) {
        return elements.stream()
                .filter(t::isInstance)
                .map(t::<T>mapInstance)
                .collect(Collectors.toList());
    }

    public static List<Cipherable> getAll() {
        return elements;
    }

    private static File getFile() {
        if (file == null) {
            String url = Main.class.getResource("/eu/shooktea/passkeeper/Main.class").toString();
            if (url.startsWith("jar:file:")) {
                url = url.substring("jar:file:".length()).replaceAll("!.*$", "");
            }
            else {
                url = url.substring("file:".length());
            }
            File classOrJarFile = new File(url);
            File directory = classOrJarFile.getParentFile();
            file = new File(directory, ".password_keeper");
        }
        return file;
    }

    private static List<Cipherable> elements = new ArrayList<>();
    private static int editedIndex = -1;
    private static File file = null;
}
